File: python/Cucumber/CucumberExpressions/cucumber_expression.py
from typing import Optional

from Cucumber.CucumberExpressions.argument import Argument
from Cucumber.CucumberExpressions.ast import Node, NodeType
from Cucumber.CucumberExpressions.cucumber_expression_parser import (
    CucumberExpressionParser,
)
from Cucumber.CucumberExpressions.parameter_type import ParameterType
from Cucumber.CucumberExpressions.tree_regexp import TreeRegexp
from Cucumber.CucumberExpressions.exceptions.errors import (
    UndefinedParameterTypeError,
    ParameterIsNotAllowedInOptional,
    OptionalIsNotAllowedInOptional,
    OptionalMayNotBeEmpty,
    AlternativeMayNotBeEmpty,
    AlternativeMayNotExclusivelyContainOptionals,
)

ESCAPE_PATTERN = b"([\\^\[({$.|?*+})\]])"  # noqa W605


class CucumberExpression:
    def __init__(self, expression, parameter_type_registry):
        self.expression = expression
        self.parameter_type_registry = parameter_type_registry
        self.parameter_types: list[ParameterType] = []
        parser = CucumberExpressionParser()
        _ast = parser.parse(expression)
        pattern = self.rewrite_to_regex(_ast)
        self.tree_regexp = TreeRegexp(pattern)

    def match(self, text: str) -> Optional[list[Argument]]:
        return Argument.build(self.tree_regexp, text, self.parameter_types)

    @property
    def source(self):
        return self.expression

    @property
    def regexp(self):
        return self.tree_regexp.regexp

    def __str__(self):
        self.source()

    def rewrite_to_regex(self, node: Node):
        if node.ast_type == NodeType.TEXT:
            return self.escape_regex(node.text)
        elif node.ast_type == NodeType.OPTIONAL:
            return self.rewrite_optional(node)
        elif node.ast_type == NodeType.ALTERNATION:
            return self.rewrite_alternation(node)
        elif node.ast_type == NodeType.ALTERNATIVE:
            return self.rewrite_alternative(node)
        elif node.ast_type == NodeType.PARAMETER:
            return self.rewrite_parameter(node)
        elif node.ast_type == NodeType.EXPRESSION:
            return self.rewrite_expression(node)
        else:
            # Can't happen as long as the switch case is exhaustive
            raise Exception(node.ast_type)

    @staticmethod
    def escape_regex(expression):
        return expression.translate({i: "\\" + chr(i) for i in ESCAPE_PATTERN})

    def rewrite_optional(self, node: Node):
        if _possible_node_with_params := self.get_possible_node_with_parameters(node):
            raise ParameterIsNotAllowedInOptional(
                _possible_node_with_params, self.expression
            )
        if _possible_node_with_optionals := self.get_possible_node_with_optionals(node):
            raise OptionalIsNotAllowedInOptional(
                _possible_node_with_optionals, self.expression
            )
        if self.are_nodes_empty(node):
            raise OptionalMayNotBeEmpty(node, self.expression)
        regex = "".join([self.rewrite_to_regex(_node) for _node in node.nodes])
        return f"(?:{regex})?"

    def rewrite_alternation(self, node: Node):
        for alternative in node.nodes:
            if not len(alternative.nodes):
                raise AlternativeMayNotBeEmpty(alternative, self.expression)
            if self.are_nodes_empty(alternative):
                raise AlternativeMayNotExclusivelyContainOptionals(
                    alternative, self.expression
                )
        regex = "|".join([self.rewrite_to_regex(_node) for _node in node.nodes])
        return f"(?:{regex})"

    def rewrite_alternative(self, node: Node):
        return "".join([self.rewrite_to_regex(_node) for _node in node.nodes])

    def rewrite_parameter(self, node: Node):
        name = node.text
        parameter_type = self.parameter_type_registry.lookup_by_type_name(name)

        if not parameter_type:
            raise UndefinedParameterTypeError(node, self.expression, name)

        self.parameter_types.append(parameter_type)

        regexps = parameter_type.regexps
        if len(regexps) == 1:
            return f"({regexps[0]})"
        return f"((?:{')|(?:'.join(regexps)}))"

    def rewrite_expression(self, node: Node):
        regex = "".join([self.rewrite_to_regex(_node) for _node in node.nodes])
        return f"^{regex}$"

    @staticmethod
    def are_nodes_empty(node: Node) -> bool:
        return not bool(
            [ast_node for ast_node in node.nodes if ast_node.ast_type == NodeType.TEXT]
        )

    @staticmethod
    def get_possible_node_with_parameters(node: Node) -> Optional[Node]:
        results = [
            ast_node
            for ast_node in node.nodes
            if ast_node.ast_type == NodeType.PARAMETER
        ]
        return results[0] if results else None

    @staticmethod
    def get_possible_node_with_optionals(node: Node) -> Optional[Node]:
        results = [
            ast_node
            for ast_node in node.nodes
            if ast_node.ast_type == NodeType.OPTIONAL
        ]
        return results[0] if results else None

File: elixir/lib/cucumber/cucumber_expressions/cucumber_expression.ex
defmodule Cucumber.CucumberExpressions.CucumberExpression do
  @moduledoc """
  A compiled Cucumber Expression.

  Created with `compile/2` (or `compile!/2`) against a
  `Cucumber.CucumberExpressions.ParameterTypeRegistry`; match text with `match/2`.
  """

  alias Cucumber.CucumberExpressions.{
    Argument,
    CucumberExpressionParser,
    Error,
    Node,
    ParameterTypeRegistry,
    TreeRegexp,
    UndefinedParameterTypeError
  }

  @enforce_keys [:source, :tree_regexp, :parameter_types]
  defstruct [:source, :tree_regexp, :parameter_types]

  @type t :: %__MODULE__{
          source: String.t(),
          tree_regexp: TreeRegexp.t(),
          parameter_types: [Cucumber.CucumberExpressions.ParameterType.t()]
        }

  @escape_pattern ~r/([\\^\[({$.|?*+})\]])/

  @spec compile(String.t(), ParameterTypeRegistry.t()) ::
          {:ok, t()} | {:error, Error.t() | UndefinedParameterTypeError.t()}
  def compile(expression, %ParameterTypeRegistry{} = registry) do
    with {:ok, ast} <- CucumberExpressionParser.parse(expression),
         {:ok, pattern, parameter_types} <- rewrite_to_regex(ast, expression, registry) do
      {:ok,
       %__MODULE__{
         source: expression,
         tree_regexp: TreeRegexp.new!(pattern),
         parameter_types: parameter_types
       }}
    end
  end

  @spec compile!(String.t(), ParameterTypeRegistry.t()) :: t()
  def compile!(expression, %ParameterTypeRegistry{} = registry) do
    case compile(expression, registry) do
      {:ok, compiled} -> compiled
      {:error, error} -> raise error
    end
  end

  @doc """
  Matches `text` against the expression. Returns a list of
  `Cucumber.CucumberExpressions.Argument`s (get each value with
  `Argument.value/1`), or `nil` if the text does not match.
  """
  @spec match(t(), String.t()) :: [Argument.t()] | nil
  def match(%__MODULE__{tree_regexp: tree_regexp, parameter_types: parameter_types}, text) do
    Argument.build(tree_regexp, text, parameter_types)
  end

  @doc "The compiled `Regex` for this expression."
  @spec regexp(t()) :: Regex.t()
  def regexp(%__MODULE__{tree_regexp: tree_regexp}), do: tree_regexp.regexp

  # Rewrites an AST node to a regex source string, accumulating the parameter
  # types referenced by the expression in match order.
  defp rewrite_to_regex(ast, expression, registry) do
    with {:ok, pattern, reversed_parameter_types} <- rewrite(ast, expression, registry, []) do
      {:ok, pattern, Enum.reverse(reversed_parameter_types)}
    end
  end

  defp rewrite(%Node{type: :text_node, token: token}, _expression, _registry, acc) do
    {:ok, process_escapes(token), acc}
  end

  defp rewrite(%Node{type: :optional_node} = node, expression, registry, acc) do
    with :ok <- assert_no_parameters(node, expression),
         :ok <- assert_no_optionals(node, expression),
         :ok <- assert_not_empty(node, expression, &Error.optional_may_not_be_empty/2) do
      rewrite_nodes(node.nodes, "", "(?:", ")?", expression, registry, acc)
    end
  end

  defp rewrite(%Node{type: :alternation_node} = node, expression, registry, acc) do
    with :ok <- assert_alternatives_valid(node, expression) do
      rewrite_nodes(node.nodes, "|", "(?:", ")", expression, registry, acc)
    end
  end

  defp rewrite(%Node{type: :alternative_node} = node, expression, registry, acc) do
    rewrite_nodes(node.nodes, "", "", "", expression, registry, acc)
  end

  defp rewrite(%Node{type: :parameter_node} = node, expression, registry, acc) do
    type_name = Node.text(node)

    case ParameterTypeRegistry.lookup_by_type_name(registry, type_name) do
      nil ->
        {:error, UndefinedParameterTypeError.new(expression, node, type_name)}

      parameter_type ->
        {:ok, build_capture_regexp(parameter_type.regexps), [parameter_type | acc]}
    end
  end

  defp rewrite(%Node{type: :expression_node} = node, expression, registry, acc) do
    rewrite_nodes(node.nodes, "", "^", "$", expression, registry, acc)
  end

  defp rewrite_nodes(nodes, delimiter, prefix, suffix, expression, registry, acc) do
    result =
      Enum.reduce_while(nodes, {:ok, [], acc}, fn node, {:ok, patterns, acc} ->
        case rewrite(node, expression, registry, acc) do
          {:ok, pattern, acc} -> {:cont, {:ok, [pattern | patterns], acc}}
          {:error, _} = error -> {:halt, error}
        end
      end)

    with {:ok, patterns, acc} <- result do
      {:ok, prefix <> Enum.join(Enum.reverse(patterns), delimiter) <> suffix, acc}
    end
  end

  defp build_capture_regexp([regexp]), do: "(" <> regexp <> ")"

  defp build_capture_regexp(regexps) do
    "(" <> Enum.map_join(regexps, "|", &"(?:#{&1})") <> ")"
  end

  defp process_escapes(text), do: Regex.replace(@escape_pattern, text, "\\\\\\1")

  defp assert_alternatives_valid(%Node{nodes: alternatives}, expression) do
    Enum.reduce_while(alternatives, :ok, fn alternative, :ok ->
      cond do
        alternative.nodes == [] ->
          {:halt, {:error, Error.alternative_may_not_be_empty(expression, alternative)}}

        not Enum.any?(alternative.nodes, &(&1.type == :text_node)) ->
          {:halt,
           {:error,
            Error.alternative_may_not_exclusively_contain_optionals(expression, alternative)}}

        true ->
          {:cont, :ok}
      end
    end)
  end

  defp assert_not_empty(%Node{nodes: nodes} = node, expression, build_error) do
    if Enum.any?(nodes, &(&1.type == :text_node)) do
      :ok
    else
      {:error, build_error.(expression, node)}
    end
  end

  defp assert_no_parameters(%Node{nodes: nodes}, expression) do
    case Enum.find(nodes, &(&1.type == :parameter_node)) do
      nil -> :ok
      child -> {:error, Error.parameter_is_not_allowed_in_optional(expression, child)}
    end
  end

  defp assert_no_optionals(%Node{nodes: nodes}, expression) do
    case Enum.find(nodes, &(&1.type == :optional_node)) do
      nil -> :ok
      child -> {:error, Error.optional_is_not_allowed_in_optional(expression, child)}
    end
  end
end
